File: code/ch09/java/ApplyDiscountTest.java
// §9.2.1 — the flagship apply_discount unit-test suite, JUnit 5 variant.
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

class ApplyDiscountTest {
  /** Reduce price by percent (0..100). Throws IllegalArgumentException on bad input. */
  static double applyDiscount(double price, double percent) {
    if (price < 0) throw new IllegalArgumentException("price must be non-negative");
    if (percent < 0 || percent > 100)
      throw new IllegalArgumentException("percent must be in 0..100");
    return Math.round(price * (1 - percent / 100) * 100) / 100.0;
  }

  @Test void noDiscount()    { assertEquals(100.0, applyDiscount(100.0, 0)); }
  @Test void halfOff()       { assertEquals(50.0, applyDiscount(100.0, 50)); }
  @Test void roundsToCents() { assertEquals(8.99, applyDiscount(9.99, 10)); }
  @Test void fullDiscount()  { assertEquals(0.0, applyDiscount(40.0, 100)); }
  @Test void freeItemAllowed() {
    assertEquals(0.0, applyDiscount(0.0, 50));  // kills the `price <= 0` mutant
  }
  @Test void rejectsBadPercent() {
    assertThrows(IllegalArgumentException.class, () -> applyDiscount(100.0, 150));
  }
}

File: code/ch09/javascript/apply_discount.test.js
// §9.2.1 — the flagship apply_discount unit-test suite, node:test variant.
const test = require("node:test");
const assert = require("node:assert/strict");

// Reduce price by percent (0..100). Throws RangeError on bad input.
function applyDiscount(price, percent) {
  if (price < 0) throw new RangeError("price must be non-negative");
  if (percent < 0 || percent > 100) throw new RangeError("percent must be in 0..100");
  return Math.round(price * (1 - percent / 100) * 100) / 100;
}

test("no discount", () => assert.equal(applyDiscount(100.0, 0), 100.0));
test("half off", () => assert.equal(applyDiscount(100.0, 50), 50.0));
test("rounds to cents", () => assert.equal(applyDiscount(9.99, 10), 8.99));
test("full discount", () => assert.equal(applyDiscount(40.0, 100), 0.0));
test("rejects bad percent", () => {
  assert.throws(() => applyDiscount(100.0, 150), RangeError);
});
test("free item allowed", () => {
  assert.equal(applyDiscount(0.0, 50), 0.0);  // kills the `price <= 0` mutant
});
